lib/test-utils/src/Test/Hspec/Extra.hs

```diff
@@ -104,17 +113,38 @@ aroundAll acquire =
 -- | A drop-in replacement for 'it' that'll automatically retry a scenario once
 -- if it fails, to cope with potentially flaky tests.
 it :: HasCallStack => String -> ActionWith ctx -> SpecWith ctx
-it title action = specify title $ \ctx -> do
-   res1 <- try $ action ctx
-   case res1 of
-       Right r1 -> return r1
-       Left (e1 :: SomeException) -> do
-           res2 <- try $ action ctx
-           case res2 of
-               -- If the second try fails, return the first error. The
-               -- second error might not be helpful.
-               Left (_e2 :: SomeException) -> throwIO e1
-               Right r2 -> return r2
+it = itWithCustomTimeout (10*minute)
+  where
+    minute = 60*1000*1000
+
+-- | Like @it@ but with a custom timeout, which makes it realistic to test.
+itWithCustomTimeout :: HasCallStack => Int -> String -> ActionWith ctx -> SpecWith ctx
+itWithCustomTimeout micro title action = specify title $ \ctx -> do
+    race_ timeout $ do
+        res1 <- try $ action ctx
+        case res1 of
+            Right r1 -> return r1
+            Left (e1 :: SomeException)
+                -- If we time out during the first attempt, then we /should/ time
+                -- out.
+                | isAsyncCancelled e1 -> throwIO e1
+                | otherwise           -> do
+                res2 <- try $ action ctx
+                case res2 of
+                    -- If the second try fails, return the first error. The
+                    -- second error might not be helpful.
+                    Left (e2 :: SomeException)
+                        | isAsyncCancelled e2 -> throwIO e1
+                        | otherwise           -> throwIO e1
+                    Right r2 -> return r2
+  where
+    isAsyncCancelled = isJust . fromException @AsyncCancelled
```

Comment:
Without this, the `race` cancel is ignored and the timeout doesn't work. Good that I tested it.